### PR TITLE
wgsl: Implement AF -> AF assignment tests

### LINKED ISSUE #2755: wgsl: execution tests for assigning `AbstractFloat` to `AbstractFloat`
This should always be precise.
This will provide coverage for the infrastructure related to extracting AbstractFloats from shaders

### PR DESCRIPTION
**Pull request by zoddicus**
This adds in the core infrastructure for running AbstractFloat returrning tests. f64 returning tests are forbidden in a few more places. A inline snippet for extracting a AbstractFloat as two u32s is used.

Fixes #2755

<hr>

**Requirements for PR author:**

- [x] All missing test coverage is tracked with "TODO" or `.unimplemented()`.
- [x] New helpers are `/** documented */` and new helper files are found in `helper_index.txt`.
- [x] Test behaves as expected in a WebGPU implementation. (If not passing, explain above.)

**Requirements for [reviewer sign-off](https://github.com/gpuweb/cts/blob/main/docs/reviews.md):**

- [ ] Tests are properly located in the test tree.
- [ ] [Test descriptions](https://github.com/gpuweb/cts/blob/main/docs/intro/plans.md) allow a reader to "read only the test plans and evaluate coverage completeness", and accurately reflect the test code.
- [ ] Tests provide complete coverage (including validation control cases). **Missing coverage MUST be covered by TODOs.**
- [ ] Helpers and types promote readability and maintainability.

When landing this PR, be sure to make any necessary issue status updates.

### CODE REVIEW

File: src/webgpu/shader/execution/expression/expression.ts [hunk 1]
```diff
@@ -496,14 +507,86 @@ export function basicExpressionBuilder(expressionBuilder: ExpressionBuilder): Sh
...
+            (c, i) => `  {
+    const exponent_bias = 1022;
+    const f = frexp(abs(${expressionBuilder(map(c.input, v => v.wgsl()))}));
+    const is_zero : bool = (f.exp == 0) && (f.fract == 0.0);
+    const_assert(is_zero || (f.exp != 0)); // Subnormals not currently supported, but zero is
+    const sign_bit : u32 = select(0, 0x80000000, ${expressionBuilder(map(c.input, v => v.wgsl()))} < 0);
+    const exponent_bits : u32 = select(u32(f.exp + exponent_bias) << 20, 0, is_zero);
+    const high_mantissa_bits : u32 = select(u32(f.fract * (1 << 21)) & ((1 << 20) - 1), 0, is_zero);
+    const low_mantissa = f.fract - floor(f.fract * (1 << 20)) / (1 << 20);
+    const adjusted_low_mantissa_bits = low_mantissa * (1 << 53) - 0x100000000;
```
Comment (zoddicus) [inline src/webgpu/shader/execution/expression/expression.ts:582]: The fact that I have to select between adjust/not adjust here is the part that I wasn't expecting/would like someone to take a look at.

I will provide some examples of why this is needed in  a comment on this PR

**Comment by zoddicus** [comment]
An example where a `- 0x100000000;` causes a negative value
```
@compute @workgroup_size(1)
fn main() {
  {
    const exponent_bias = 1022;
    const f = frexp(abs(3.4615312390232186e+34));
    const is_zero : bool = (f.exp == 0) && (f.fract == 0.0);
    const_assert(is_zero || (f.exp != 0)); // Subnormals not currently supported, but zero is
    const sign_bit : u32 = select(0, 0x80000000, 3.4615312390232186e+34 < 0);
    const exponent_bits : u32 = select(u32(f.exp + exponent_bias) << 20, 0, is_zero);
    const high_mantissa_bits : u32 = select(u32(f.fract * (1 << 21)) & ((1 << 20) - 1), 0, is_zero);
    const low_mantissa = f.fract - floor(f.fract * (1 << 20)) / (1 << 20);
    const adjusted_low_mantissa_bits = low_mantissa * (1 << 53) - 0x100000000;
    // const unadjusted_low_mantissa_bits = low_mantissa * (1 << 53);
    // const low_mantissa_bits = u32(select(adjusted_low_mantissa_bits, unadjusted_low_mantissa_bits, adjusted_low_mantissa_bits  < 0));
    const low_mantissa_bits = u32(adjusted_low_mantissa_bits);
    outputs[0].value.x = select(low_mantissa_bits, 0, is_zero);
    outputs[0].value.y = sign_bit | exponent_bits | high_mantissa_bits;
  }
}warning:
1 error(s) generated while compiling the shader:
:26:31 error: value -1431655765.0 cannot be represented as 'u32'
    const low_mantissa_bits = u32(adjusted_low_mantissa_bits);
                              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

An example where removing `- 0x100000000;` causes a larger than u32 value
```
@compute @workgroup_size(1)
fn main() {
  {
    const exponent_bias = 1022;
    const f = frexp(abs(2.576358348678172e+171));
    const is_zero : bool = (f.exp == 0) && (f.fract == 0.0);
    const_assert(is_zero || (f.exp != 0)); // Subnormals not currently supported, but zero is
    const sign_bit : u32 = select(0, 0x80000000, 2.576358348678172e+171 < 0);
    const exponent_bits : u32 = select(u32(f.exp + exponent_bias) << 20, 0, is_zero);
    const high_mantissa_bits : u32 = select(u32(f.fract * (1 << 21)) & ((1 << 20) - 1), 0, is_zero);
    const low_mantissa = f.fract - floor(f.fract * (1 << 20)) / (1 << 20);
    // const adjusted_low_mantissa_bits = low_mantissa * (1 << 53) - 0x100000000;
    const unadjusted_low_mantissa_bits = low_mantissa * (1 << 53);
    // const low_mantissa_bits = u32(select(adjusted_low_mantissa_bits, unadjusted_low_mantissa_bits, adjusted_low_mantissa_bits  < 0));
    const low_mantissa_bits = u32(unadjusted_low_mantissa_bits);
    outputs[0].value.x = select(low_mantissa_bits, 0, is_zero);
    outputs[0].value.y = sign_bit | exponent_bits | high_mantissa_bits;
  }
}warning:
1 error(s) generated while compiling the shader:
:26:31 error: value 5726623061.0 cannot be represented as 'u32'
    const low_mantissa_bits = u32(unadjusted_low_mantissa_bits);
                              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

File: src/webgpu/shader/execution/expression/expression.ts [hunk 2]
```diff
@@ -496,14 +507,86 @@ export function basicExpressionBuilder(expressionBuilder: ExpressionBuilder): Sh
...
+        // Documentation of the snippet out of line, since shader length affects
+        // compilation time significantly on some backends.
+        //
+        //   const exponent_bias = 1022;
+        //   // Use frexp() to obtain the exponent and fractional parts
+        //   const f = frexp(abs(${expressionBuilder(map(c.input, v => v.wgsl()))}));
+        //
+        //   // Zero behaves like a subnormal, so needs special casing later
+        //   const is_zero : bool = (f.exp == 0) && (f.fract == 0.0);
+        //   const_assert(is_zero || (f.exp != 0)); // Subnormals not currently supported, but zero is
```
Comment (dneto0) [inline src/webgpu/shader/execution/expression/expression.ts:540]: Please make the "Subnormals not supported" fact more prominent.  Put it near the top, near 521

File: src/webgpu/shader/execution/expression/expression.ts [hunk 3]
```diff
@@ -496,14 +507,86 @@ export function basicExpressionBuilder(expressionBuilder: ExpressionBuilder): Sh
...
+        //
+        //   // MSB of the high_u32 is 1 if the value is negative, otherwise 0
+        //   const sign_bit : u32 = select(0, 0x80000000, ${expressionBuilder(map(c.input, v => v.wgsl()))} < 0);
+        //
+        //   // The exponent is stored in bits [20..31] of high_u32, special case for zero
+        //   const exponent_bits : u32 = select(u32(f.exp + exponent_bias) << 20, 0, is_zero);
+        //
+        //   // Extract the high 20 bits of the mantissa as u32, special case for zero
+        //   const high_mantissa_bits : u32 = select(u32(f.fract * (1 << 21)) & ((1 << 20) - 1), 0, is_zero);
```
Comment (dneto0) [inline src/webgpu/shader/execution/expression/expression.ts:549]: When the original number is normal, the use of 21 here makes sense:  It shifts the implied leading 1 mantissa bit out of the way and throws it away.

When the original number is subnormal this throws away the bit.  

The method is not documented as ignoring subnormals.

File: src/webgpu/shader/execution/expression/expression.ts [hunk 4]
```diff
@@ -496,14 +507,86 @@ export function basicExpressionBuilder(expressionBuilder: ExpressionBuilder): Sh
...
+        //
+        //   // Subtract out the high 20 bits from the mantissa.
+        //   // const low_mantissa = f.fract - floor(f.fract * (1 << 20)) / (1 << 20);
```
Comment (dneto0) [inline src/webgpu/shader/execution/expression/expression.ts:552]: Why is this using 1 << 20 when the upper bits are extracted with 1 << 21 ?


This code could use an example value as an illustration.  e.g. 1.9abcdefp+1

**Review by dneto0 — CHANGES_REQUESTED** [review]
I find this confusing so far, so I am not confident to review the bit you wanted reviewed.

File: src/webgpu/shader/execution/expression/expression.ts (hunk 4 above)
Comment (dneto0) [reply src/webgpu/shader/execution/expression/expression.ts:552]: Also, I recommend adding unit tests.

**Review by zoddicus — COMMENTED** [review]
PTAL, I have responded to review comments, added FTZ behaviour, cleanedup documentation, and added explicit testing.

File: src/webgpu/shader/execution/expression/expression.ts [hunk 5]
```diff
@@ -496,14 +513,113 @@ export function basicExpressionBuilder(expressionBuilder: ExpressionBuilder): Sh
...
+        //
+        //   // Extract the high 20 bits of the mantissa as u32, masking off
+        //   // unneeded bits.
+        //   const high_mantissa_bits : u32 = (u32(f_fract * (1 << 20))) & 0x000fffff;
+        //
+        //   // Subtract out the high bits from the mantissa.
+        //   // const low_mantissa = f_fract - floor(f_fract * (1<< 21)) / (1 << 21);
+        //
+        //   // Extract the low 32 bits of the mantissa as an u32
+        //   // The calculated lower mantissa may or may not have a leading 1 in
```
Comment (dneto0) [inline src/webgpu/shader/execution/expression/expression.ts:585]: I can't figure out why that leading 1 should be there, except by a numerical error in the calculation somewhere along the way.

The division `  / (1 << 21)` looks fishy to me. See below.

File: src/webgpu/shader/execution/expression/expression.ts [hunk 6]
```diff
@@ -496,14 +513,113 @@ export function basicExpressionBuilder(expressionBuilder: ExpressionBuilder): Sh
...
+        //   // of range. If the bit isn't present, but removed without checking
+        //   // for its presences, the result will be negative and the wrong
+        //   // magnitude.
+        //   // So when converting to u32, the value needs to be selected on to
+        //   // remove that bit iff it is present.
+        //   const low_mantissa_bits = (low_mantissa * (1 << 52));
+        //
+        //   // Pack the result into 2 u32s, adjusting the low mantissa if
+        //   // needed
+        //   outputs[${i}].value.x = u32(select(low_mantissa_bits, low_mantissa_bits - 0x100000000, low_mantissa_bits >= 0x100000000));
```
Comment (dneto0) [inline src/webgpu/shader/execution/expression/expression.ts:601]: The only reason I can think that there is a 1 in the 33rd position is if there is a numerical error in the computation somewhere. Maybe it's in the division `1 / (1<<21)`


I recoded the algorithm using ldexp instead and it's much more straightforward, I think.  

This uses C++17 so I can use hex floats.  (And I assume `unsigned` is the same as `uint32_t`, which is common these days.)

```cpp
#include <iostream>
#include <iomanip>
#include <cmath>

#define SHOW(X)  std::cout << std::setw(15) << " " #X << "  " << std::hex << std::hexfloat << (X) << std::endl

int main() {

  const double orig = 0x1.921fb54442d18p+1; // This is PI
  int n;
  const auto f = frexp(orig,&n);
  SHOW(f);

  const auto highf = ldexp(f,21); // 21 is the 20 bits we want plus the implicit leading 1
  SHOW(highf);
  const auto high = unsigned(highf);
  SHOW(high);

  const auto lowf = f - ldexp(floor(highf),-21);
  SHOW(lowf);
  const auto low = unsigned(ldexp(lowf,53));   // 53 = 32 + 21
  SHOW(low);
}
```
And here is the output:

```
$ ./a
              f  0x1.921fb54442d18p-1
          highf  0x1.921fb54442d18p+20
           high  1921fb
           lowf  0x1.5110b46p-23
            low  54442d18
```

Note that 'high' still has the 21st bit set, so you do have to mask it off.

They key is we recover exactly the right bits, unchanged.

**Review by dneto0 — CHANGES_REQUESTED** [review]
I still find the algorithm mysterious in spots.

I've offered a simplification to key parts.

File: src/webgpu/shader/execution/expression/expression.ts [hunk 7]
```diff
@@ -496,14 +513,113 @@ export function basicExpressionBuilder(expressionBuilder: ExpressionBuilder): Sh
...
+        //   const f_fract = select(f.fract, 0, subnormal_or_zero);
+        //   const f_exp = select(f.exp, -kExponentBias, subnormal_or_zero);
+        //
+        //   // The exponent is stored in bits [20..31] of high_u32, special
+        //   // case for zero.
+        //   const exponent_bits : u32 = (f_exp + kExponentBias) << 20;
+        //
+        //   // Extract the high 20 bits of the mantissa as u32, masking off
+        //   // unneeded bits.
+        //   const high_mantissa_bits : u32 = (u32(f_fract * (1 << 20))) & 0x000fffff;
```
[the remark this answers is not in the document]
Comment (zoddicus) [reply src/webgpu/shader/execution/expression/expression.ts]: Done

File: src/webgpu/shader/execution/expression/expression.ts (hunk 5 above)
Comment (zoddicus) [reply src/webgpu/shader/execution/expression/expression.ts]: I agree

File: src/webgpu/shader/execution/expression/expression.ts [hunk 8]
```diff
@@ -496,14 +513,113 @@ export function basicExpressionBuilder(expressionBuilder: ExpressionBuilder): Sh
...
+        //   // cannot be directly converted to an AbstractInt, so there is no
+        //   // simple way to just mask the bits that are needed without
+        //   // concretizing first.
+        //   // If that bit is present the u32 cast will fail due to being out
+        //   // of range. If the bit isn't present, but removed without checking
+        //   // for its presences, the result will be negative and the wrong
+        //   // magnitude.
+        //   // So when converting to u32, the value needs to be selected on to
+        //   // remove that bit iff it is present.
+        //   const low_mantissa_bits = (low_mantissa * (1 << 52));
```
[the remark this answers is not in the document]
Comment (zoddicus) [reply src/webgpu/shader/execution/expression/expression.ts]: Done

File: src/webgpu/shader/execution/expression/expression.ts (hunk 6 above)
Comment (zoddicus) [reply src/webgpu/shader/execution/expression/expression.ts]: Done

**Review by zoddicus — COMMENTED** [review]
Reworked implementation and documentation based on your snippet, it is much more readable now.

File: src/webgpu/shader/execution/expression/expression.ts [hunk 9]
```diff
@@ -496,14 +513,111 @@ export function basicExpressionBuilder(expressionBuilder: ExpressionBuilder): Sh
...
+        //   // [20..31] of the upper u32
+        //   const exponent_bits : u32 = (f_exp + kExponentBias) << 20;
+        //
+        //   // Extract the portion of the mantissa that appears in upper u32 as
+        //   // a float for later use
+        //   const high_mantissa = ldexp(f_fract, 21);
+        //
+        //   // Extract the portion of the mantissa that appears in upper u32 as
+        //   // as bits. This value is masked, because normals will explicitly
+        //   // have the implicit leading 1 that, which shouldn't be in the
```
Comment (dneto0) [inline src/webgpu/shader/execution/expression/expression.ts:586]: nit: grammar  "have the implicit leading 1 that should not be in the final result".

**Review by dneto0 — APPROVED** [review]
Nice.  :-)

Thanks!.  Grammar nit.

File: src/webgpu/shader/execution/expression/expression.ts (hunk 9 above)
Comment (zoddicus) [reply src/webgpu/shader/execution/expression/expression.ts]: Done

**Pull request merged**